### PR TITLE
Add tracing to Android

### PR DESCRIPTION
**Pull request by renderexpert**
# Why:
Check performance progression/regression, specifically in the rendering code.

# How:
Integrate system tracing for Android by this guide: https://developer.android.com/ndk/reference/group/tracing

# Examples of usage:
Screenshot from 2024-05-25 11-35-09
Screenshot from 2024-05-25 11-35-56

### CODE REVIEW

**Comment by biodranik** [comment]
Is there a way to see what is currently rendered on textures (and which textures are used) in the emulator/on a device?

**Comment by renderexpert** [comment]
@biodranik There is a 3party tool for that called RenderDoc - https://renderdoc.org/

**Comment by renderexpert** [comment]
Update: I managed to capture Vulkan too using button "Cycle Active Window" in Renderdoc.
Screenshot from 2024-05-26 14-58-36

**Comment by Jean-BaptisteC** [comment]
@renderexpert rebase?

File: drape/vulkan/vulkan_object_manager.cpp [hunk 1]
```diff
@@ -16,6 +18,9 @@ namespace
 size_t constexpr kBackendQueueIndex = 0;
 size_t constexpr kOtherQueueIndex = 0;
 
+// Maximum descriptors sets count in the pool.
+uint32_t constexpr kMaxDescriptorsSetCount = 256 * kMaxInflightFrames;
```
Comment (biodranik) [inline drape/vulkan/vulkan_object_manager.cpp:22]: Should it be under the ifdef? @renderexpert

File: platform/trace.hpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+class TraceSection
+{
+public:
+  inline TraceSection(Trace & trace, char const * section) noexcept 
+    : m_trace(trace) 
+  {
+    m_trace.BeginSection(section);
+  }
+
+  inline ~TraceSection() noexcept {
```
Comment (biodranik) [inline platform/trace.hpp:39]:
```suggestion
  inline ~TraceSection() noexcept
  {
```

File: platform/trace_android.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+class TraceImpl
+{
+public:
+  TraceImpl() noexcept
+  {
+    m_lib = dlopen("libandroid.so", RTLD_NOW | RTLD_LOCAL);
+
+    // Access the native tracing functions.
+    if (m_lib != nullptr) {
```
Comment (biodranik) [inline platform/trace_android.cpp:21]:
```suggestion
    if (m_lib != nullptr)
    {
```

File: platform/trace_android.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      // Use dlsym() to prevent crashes on devices running Android 5.1
+      // (API level 22) or lower.
+      m_beginSection = reinterpret_cast<ATrace_beginSection>(
+          dlsym(m_lib, "ATrace_beginSection"));
+      m_endSection = reinterpret_cast<ATrace_endSection>(
+          dlsym(m_lib, "ATrace_endSection"));
+      m_setCounter = reinterpret_cast<ATrace_setCounter>(
+          dlsym(m_lib, "ATrace_setCounter"));
```
Comment (biodranik) [inline platform/trace_android.cpp:29]:
```suggestion
      m_beginSection = reinterpret_cast<ATrace_beginSection>(dlsym(m_lib, "ATrace_beginSection"));
      m_endSection = reinterpret_cast<ATrace_endSection>(dlsym(m_lib, "ATrace_endSection"));
      m_setCounter = reinterpret_cast<ATrace_setCounter>(dlsym(m_lib, "ATrace_setCounter"));
```

File: platform/trace_android.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  ATrace_beginSection m_beginSection = nullptr;
+  ATrace_endSection m_endSection = nullptr;
+  ATrace_setCounter m_setCounter = nullptr;
+};
+
+Trace::Trace() noexcept 
+  : m_impl(std::make_unique<TraceImpl>())
+{}
+
+Trace::~Trace() noexcept = default;
```
Comment (vng) [inline platform/trace_android.cpp:66]: Formally, ctor and dtor here are not ```noexcept```

File: platform/platform.hpp [hunk 6]
```diff
@@ -118,6 +119,8 @@ class Platform
 
   platform::BatteryLevelTracker m_batteryTracker;
 
+  platform::Trace m_trace;
```
Comment (vng) [inline platform/platform.hpp:122]: Isn't it better to make it like a singleton ```platform::Trace & Trace::instance()``` without Platform dependency?

File: platform/platform.hpp [hunk 7]
```diff
@@ -4,6 +4,7 @@
 #include "platform/country_defines.hpp"
 #include "platform/gui_thread.hpp"
 #include "platform/secure_storage.hpp"
+#include "platform/trace.hpp"
```
Comment (vng) [inline platform/platform.hpp:7]: Please, remove.

File: drape_frontend/postprocess_renderer.cpp [hunk 8]
```diff
@@ -10,6 +10,8 @@
 #include "drape/texture_manager.hpp"
 #include "drape/render_state.hpp"
 
+#include "platform/platform.hpp"
```
Comment (vng) [inline drape_frontend/postprocess_renderer.cpp:13]: Replace with platform/trace.hpp in other places.

File: platform/trace.hpp [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  inline ~TraceSection() noexcept
+  {
+    Trace::Instance().EndSection();
+  }
+};
+}  // namespace platform
+
+#ifdef ENABLE_TRACE
```
Comment (vng) [inline platform/trace.hpp:47]: Adding ```&& OMIM_OS_ANDROID``` here allows remove trace_empty.cpp and discard the project.pbxproj changes.

**Review by vng — APPROVED** [review]

**Pull request merged**